contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc

```diff
@@ -144,6 +158,165 @@ Refer to:
+
+* ThreadSanitizer (TSan) detects data races. It's available in Clang and gcc.
+
+* LeakSanitizer (LSan) is a memory leak detector. It's integrated into AddressSanitizer, but there's also a standalone version that can be used with other tools.
+
+=== Cppcheck
+
+http://cppcheck.net/[Cppcheck], versions of which also run on Windows, Linux, and MacOS, is a static analysis tool for pass:[C++] code. It doesn't require the code to be executed, so it can catch bugs at compile-time that might be missed at runtime. It can check for memory leaks, null pointer dereferences, syntax errors, and other issues.
+
+== Improve CI Performance
+
+This section provides some insight into how you might use precompiled headers (PCHs), https://ccache.dev/[CCache], or https://docs.docker.com/get-docker/[Docker] to reduce continuous integration times.
+
+=== Precompiled Headers
+
+Precompiled headers can help reduce build times by precompiling certain parts of your codebase that are unlikely to change between builds, such as header files from libraries you're using or project-wide header files. Here's a basic example of how to set it up in a CMake project:
+
+[source, txt]
+----
+# In your CMakeLists.txt
+target_precompile_headers(your_project PRIVATE pch.h)
+----
+
+Then, in `pch.h`, you'd include headers that are used throughout your project and don't change often. Note that the usage of precompiled headers is highly dependent on the build system you're using (CMake, Meson, etc.) and the compiler.
+
+=== CCache
+
+https://ccache.dev/[CCache] is a compiler cache. It speeds up recompilation of pass:[C++] code by caching previous compilations and reusing them when the same compilation is done again.
+
+Here's a simple example of how to use it in a CI environment:
+
+. In your CI configuration file, add a step to install CCache. This will depend on the specific OS you're using in your CI environment. For Ubuntu or Debian, you can use:
++
+[source,txt]
+----
+sudo apt-get install ccache
+----
+
+. You'll need to prepend ccache to your compiler commands to use the cache. In a CMake-based project, you might do something like this in your CI configuration file:
++
+[source,txt]
+----
+export CXX="ccache g++" CC="ccache gcc"
+cmake ..
+make
+----
+
+. You may want to adjust cache settings depending on your project's needs. For instance, you might set `max_size` to control the maximum size of the cache:
++
+[source,txt]
+----
+ccache --max-size=2G
+----
++
+Or set sloppiness to control what aspects of the build can vary while still using the cache:
++
+[source,txt]
+----
+ccache --set-config=sloppiness=pch_defines,time_macros
+----
+
+=== Docker
+
+https://docs.docker.com/get-docker/[Docker] can be used to potentially reduce CI times, but it's important to understand that Docker isn't inherently a performance-boosting tool. Rather, it provides consistency and isolation for your applications, which indirectly may lead to improved CI times. Here are some ways Docker can help:
```

Comment:
I think Sam wouldn't agree with that. Although Ruben would be more favorable to docker in CI. I think we need to list the cases where docker is useful.

Reducing CI times is questionable in most simpler libraries. The workflow usually starts with an image and then downloading a few binaries or apt-get or something. Let's say you need GCC11 as a workflow. 

- As apt-get already gives you compiled GCC11 binaries and any binaries in the image also need to be downloaded (no free lunch), you might end up saving nothing. 
- Now consider the case where the github image already contains GCC11. Now you're actually losing a free-lunch, because you have an image with GCC11 and you are using it to download another image with GCC11.
- Then there's the problem of testing multiple compilers. If you have one docker container for each compiler, then this has a high cost. If you have one docker container with all compilers, now you have all workflows downloading all compilers and then just testing one of them, while the github image would already have most of these compilers.

Consistency and isolation are good features of docker more than reducing CI times, but it also has problems. It goes somewhat against the purpose of one of our missions which is to support the most common environments instead of an isolated environment. The scenario where this is useful is more specific and not the general case and that has to be clear in the text. There are basically two scenarios where isolation is very useful. 

- The first is when we are replicating environments we couldn't replicate otherwise. For instance, we need ubuntu 16 all the time but there's no github image for it. So we use the ubuntu 22 image and run a ubuntu 16 docker container on it. This is the most common case by far.
- The second case where it's very useful is when the workflow depends on tools that can't be easily downloaded with apt-get. One example is mrdox, where we need a very specific version of LLVM. The binaries for this version are not available anywhere and building it from source takes hours. So we have a docker container that bundles the LLVM binaries and everything is much easier. But this use case is also a little questionable because we could also store these binaries elsewhere and download them in CI, which would have the same effect. The best use case for docker seems to be really just replicating environments we couldn't replicate otherwise.